### DISCUSSION TITLE
Simplify HMR setups for server-side Remix

### DISCUSSION THREAD

**Discussion by kettanaito** [remix-run/remix#6556] [Proposals]
## Context

I want to have a custom logic applied process-wide server-side in Remix. I also want for that logic to persist between hot updates in Remix without duplicating or terminating itself. Note that I'm _not_ speaking about the `loader` but rather from the third-party tooling perspective (please see the example below). 

## How this can be achieved today

Right now, I have to resort to quite boilerplate-y setup in order for my logic to survive through hot updates in Remix:

```ts
declare global {
  var mswServerInstance: SetupServer
}

export function enableApiMocking(handlers: Array<RequestHandler>) {
  const existingServer = globalThis.mswServerInstance

  if (typeof existingServer !== 'undefined') {
    restart(existingServer, handlers)
  } else {
    start(setup(handlers))
  }
}

function setup(handlers: Array<RequestHandler>) {
  const server = setupServer(...handlers)
  globalThis.mswServerInstance = server
  return server
}

function start(server: SetupServer) {
  server.listen()

  process
    .once('SIGTERM', () => server.close())
    .once('SIGINT', () => server.close())
}

function restart(server: SetupServer, handlers: Array<RequestHandler>) {
  server.close()
  start(setup(handlers))
}
```

— [Full reference](https://github.com/mswjs/examples-new/blob/4035459a12f32072d6b31e75572efb6cf3c056dd/examples/with-remix/app/mocks/node.ts#L4-L35)

This is a good moment when the framework could handle this for me. 

## Proposal

Are you open to some sort of API to enable custom server-side logic across hot updates? If so, we can collaborate on one. 

There are, undoubtedly, multiple ways to do this, but I highly recommend exporting an on-demand API from some `@remix-run/*` package rather than putting this logic into configuration files. The users should remain in charge of their file structure, which includes giving them the ability to enable custom server-side logic wherever they want. 

This is roughly how I imagine such API to work:

```ts
import { setupServer } from 'msw/node'
// The new API is exported and used on-demand.
import { injectServerEffect } from '@remix-run/node'

export function enableApiMocking(handlers) {
  // The new API can be used anywhere, including within user-defined logic
  // like the "enableApiMocking" function here that is designed to be consumed
  // elsewhere.
  injectServerEffect('my-effect-id', () => {
    const server = setupServer(...handlers)
    server.listen()

    // Similar to React hooks, the new API returns a cleanup function
    // that Remix runs whenever its Node process receives an update or terminates.
    return () => {
      server.close()
    }
  })
}
```

Under the hood, such `injectServerEffect` API can be implemented like this:

```ts
// node_modules/@remix-run/node/injectServerEffect.js
export function injectServerEffect(id, effect) {
  const persistedEffect = globalThis[id]

  // If the global pointer to the effect exists, it means we are
  // in the hot update phase and need to restart the effect
  // otherwise it will become stale. 
  if (persistedEffect) {
    persistedEffect.cleanup()
    persistedEffect.run()
    return
  }

  const cleanup = effect()
  const persistedEffect = {
    run: effect,
    cleanup() {
      cleanup()
      globalThis[id] = undefined
    }
  }

  // Store the effect so Remix can flush it on next hot reload,
  // preventing it from becoming stale. 
  globalThis[id] = persistedEffect

  process.on('SIGTERM', cleanup).on('SIGINT', cleanup)
}
```

> This still has some rough edges, like annotating which data to persist across hot updates (the `server` instance in my example). 

I am open to implementing this after we discuss the API in more detail. I would also like for Remix to ship this because it's fundamentally something one needs to integrate any custom process-wide logic with Remix and its hot updates. The use cases for this are mainly for third-party libraries but it doesn't negate my point. If anything, it only makes it stronger, as the success of the Remix ecosystem is Remix's success as well. Shipping this from Remix would also make educational context (docs, tutorials) much easier to explain and consume.

**Comment by alexanderson1993** [comment]
It seems like this would be valuable for setting up Prisma instances as well. As of right now the official examples have to do a bit of global variable setting slight-of-hand to make things work properly. Without looking at the examples, a beginner might run into these kinds of foot guns.

That's not saying a beginner would know to use `insertServerEffect` either, but at least it would be an official way to accomplish the same thing.

https://github.com/remix-run/examples/blob/main/_official-jokes/app/utils/db.server.ts

**Reply by kettanaito** [reply]
The proposed API, just as the use case for it, derives mainly from third-party packages' context. I don't expect immediate Remix users to know or use `injectServerEffect`. But I'd like to have this API if I'm a library author who wants to ship great HMR experience for our shared users.

**Reply by jacobparis** [reply]
One solution in user-space is @jenseng's [singleton.ts](https://github.com/jenseng/abuse-the-platform/blob/2993a7e846c95ace693ce61626fa072174c8d9c7/app/utils/singleton.ts) that looks like this in use

@kentcdodds [uses it in the Epic Stack](https://github.com/epicweb-dev/epic-stack/blob/2c818d551b60254c2a6e914ab6dad8e76d9b887d/app/utils/db.server.ts#L4) like so
```
import { PrismaClient } from '@prisma/client'
import { singleton } from './singleton.server.ts'

const prisma = singleton('prisma', () => new PrismaClient())
prisma.$connect()

export { prisma }
```

Notably there's no cleanup handling

**Reply by kettanaito** [reply]
@jacobparis, it looks great for simple use cases! What I'd like to propose is something more universal that can handle persisted instances, like the `server` from the example above. That simply allows for more advanced side-effects for third-parties.

**Comment by kettanaito** [comment]
## Implementation draft

Here's a working implementation for the `injectServerEffect` function.

```ts
type ServerEffectFunction<Args extends Array<unknown>> = (
  ...args: Args
) => () => void

interface ServerEffect<Args extends Array<unknown>> {
  id: string
  run: ServerEffectFunction<Args>
  cleanup(): void
}

/**
 * Creates a process-wide server-side side effect that can survive
 * hot updates in Remix. 
 * @see https://github.com/remix-run/remix/discussions/6556
 */
export function injectServerEffect<Args extends Array<unknown>>(
  id: string,
  run: ServerEffectFunction<Args>,
  args: Args
) {
  const persistedEffect = globalThis[id]

  if (typeof persistedEffect !== 'undefined') {
    persistedEffect.cleanup()
    persistedEffect.cleanup = persistedEffect.run(...args)
    return
  }

  const cleanup = run(...args)
  const effect: ServerEffect<Args> = {
    id,
    run,
    cleanup,
  }

  globalThis[id] = effect

  const globalCleanup = () => {
    effect.cleanup()
    globalThis[id] = undefined
  }

  process.on('SIGTERM', globalCleanup).on('SIGINT', globalCleanup)
}
```

Interesting, but the function must accept any arguments of the effect as its dependency. Kind of reminds me of the React hooks dependencies but this one seems to have a different nature. If the effect is not passed the arguments explicitly, it will operate on stale arguments upon hot update. That happens because if you wish for a change in the argument to trigger the re-evaluation of the effect, it must be present in the import path through and through.

**Comment by kettanaito** [comment]
I've opened a pull request with the proposed changed:

- https://github.com/remix-run/remix/pull/6749

It looks like many third-party libraries, like MSW and Prisma, can take advantage of the `injectServerEffect` function. I hope this is aligned with the Remix's team vision.

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
